src/main/java/org/truffleruby/core/module/ModuleOperations.java

```diff
@@ -480,56 +499,56 @@ public static MethodLookupResult lookupSuperMethod(InternalMethod currentMethod,
         final String name = currentMethod.getSharedMethodInfo().getName(); // use the original name
         return lookupSuperMethod(
                 currentMethod.getDeclaringModule(),
+                null,
                 name,
                 objectMetaClass,
                 currentMethod.getDeclarationContext());
     }
 
     @TruffleBoundary
-    private static MethodLookupResult lookupSuperMethod(DynamicObject declaringModule, String name,
+    private static MethodLookupResult lookupSuperMethod(DynamicObject declaringModule, DynamicObject lookupTo,
+            String name,
             DynamicObject objectMetaClass, DeclarationContext declarationContext) {
         assert RubyGuards.isRubyModule(declaringModule);
         assert RubyGuards.isRubyModule(objectMetaClass);
 
-        boolean inRefinedMethod = Layouts.MODULE.getFields(declaringModule).isRefinement();
-
         final ArrayList<Assumption> assumptions = new ArrayList<>();
         boolean foundDeclaringModule = false;
-        for (DynamicObject module : Layouts.MODULE.getFields(objectMetaClass).ancestors()) {
-            if (!foundDeclaringModule) {
-                if (!inRefinedMethod) {
-                    if (module == declaringModule) {
-                        foundDeclaringModule = true;
+
+        for (DynamicObject ancestor : Layouts.MODULE.getFields(objectMetaClass).ancestors()) {
+            if (ancestor == lookupTo) {
+                return new MethodLookupResult(null, toArray(assumptions));
+            }
+
+            final DynamicObject[] refinements = getRefinementsFor(declarationContext, ancestor);
+
+            if (refinements != null) {
+                for (DynamicObject refinement : refinements) {
+                    // TODO: pass new declaration context without current refinements
+                    //       to lookup super in other namespaces
+                    final MethodLookupResult superMethodInRefinement = lookupSuperMethod(
+                            declaringModule,
+                            null, // super in refined class takes precedence
```

Comment:
Could you clarify this? I guess it relates to #2036.
So if I understand here we don't limit it, because we should look in C before other refinements?
But then if there are multiple refinements, is it even worth checking the other refinements? Or to lookup the whole ancestors again?

Reply:
> But then if there are multiple refinements, is it even worth checking the other refinements? Or to lookup the whole ancestors again?

You are right as always :) We can look in `C -> ...` only one time (the first time if we have multiple refinements). 
I decided not to make this improvement right now, because we still have to change the code because of #2036. So, I plan to do this as part of the work on the next issue. If you do not mind, of course

Reply:
> I guess the current code is not less correct than before for super?

This is the reason why I moved the bug to separate issue instead of fix it in the current scope.